### PR TITLE
Move argparse code into separate argparse modules 

### LINKED ISSUE #367: Move argparse code into separate `argparse` modules
Functions and classes handling command-line argument parsing using the `argparse` module as the base are peppered throughout the database. Functions named like `argparse_*`, and classes like `*ArgumentParser` and `*Action` reside either in `__init__.py` or `misc.py` files. If we move them to separate `argparse.py` files in each corresponding package, we could make each affected file shorter and more [cohesive](https://en.wikipedia.org/wiki/Cohesion_(computer_science)).

This is done when all listed and related functions and classes are moved to `argparse` modules in each of their packages, and all references of them are updated.

### PR DESCRIPTION
**Pull request by shivam-Purohit**
Added module argparse to separate the functions and classes using the command-line argument parsing. 
closes #367

### CODE REVIEW

**Comment by shivam-Purohit** [comment]
@spbnick Am I on the right path?  is there something I am missing?
If not then I have to repeat this for all the other files like db, oo.

**Review by spbnick — CHANGES_REQUESTED** [review]
Looks pretty good! Thank you, Shivam!

I only have a couple inline comments. Please proceed to the other modules after addressing them.

Please also observe commit guidelines in the [PR I just posted](https://github.com/kernelci/kcidb/pull/380).

Thank you :slightly_smiling_face:

File: kcidb/__init__.py
Comment (shivam-Purohit) [reply kcidb/__init__.py:238]: sure

**Comment by shivam-Purohit** [comment]
I am sorry the commits are all over the place. It will be a hard task but could you review it please @spbnick? I am also checking if I missed anything.

File: kcidb/orm/__init__.py
Comment (spbnick) [inline kcidb/orm/__init__.py:1286]: Most (if not all of these) need to stay in this module. Please only move the argparse stuff. Post questions in the issue, this PR, or on the #outreachy channel, if you have a difficulty deciding what should stay and what should move, or disentangling them. In general the approach I describe for the `db` module should be used.

**Review by spbnick — CHANGES_REQUESTED** [review]
Nice work, @shivam-Purohit!

I have a few requests inline, and please don't hesitate to ask questions, if anything is not clear, or you need help :heart:

File: kcidb/orm/__init__.py
Comment (shivam-Purohit) [reply kcidb/orm/__init__.py:1286]: I thought we will also need to move dependent functions also as the argparse functions were throwing errors. Should I import those functions that argparse_ ones require into the argparse.py ?

File: kcidb/mq/argparse.py
Comment (shivam-Purohit) [reply kcidb/mq/argparse.py:22]: In all the argparse module we would not need logger? or I need to check that

File: kcidb/db/__init__.py
Comment (spbnick) [inline kcidb/db/__init__.py:53]: Hmm, this is not how one would pass arguments to an `argparse` action. This is how you would derive from a class. You have to find out how to do this right. Perhaps passing it to the constructor (`__init__`)? Try to figure it out, but if you're stuck reach out on the Slack channel. I or somebody else will help.

Also, we need to move this action to the corresponding `argparse` module, not keep it here.

File: kcidb/mq/argparse.py
Comment (spbnick) [reply kcidb/mq/argparse.py]: You need to check if any code you're moving into the new modules needs it. If not, don't add it. Thanks!

File: kcidb/orm/__init__.py
Comment (spbnick) [reply kcidb/orm/__init__.py:1286]: The way you normally do this is by parametrizing the functions/classes being extracted from the original module, and then supplying whatever they need to them when calling/using them. Similarly to what I described for the `db` module. Let's chat on the Slack channel, or maybe have a quick call tomorrow, if this is not clear enough :slightly_smiling_face:

**Review by spbnick — CHANGES_REQUESTED** [review]
Apart from the inline comments, please pay attention to the errors that the CI checks show, and the [commit guidelines](https://github.com/kernelci/kcidb/blob/main/doc/developer_guide.md#commits).

File: kcidb/db/__init__.py
Comment (shivam-Purohit) [reply kcidb/db/__init__.py:53]: If I am passing as a parameter I should not need to import the DRIVER_TYPES. I am not finding the instance where this constructor was used so I could pass the argument. 

ubuntu  Running  - Oracle VM VirtualBox 10-03-2023 11_03_52

**Comment by shivam-Purohit** [comment]
` action= PatternHelpAction, `
should PatternHelpAction be also be moved to argparse?

File: kcidb/db/__init__.py
Comment (spbnick) [reply kcidb/db/__init__.py:53]: Right, the `*Action` constructor is called by the original `argparse` module itself. So, we cannot pass a parameter this way.

Let's try another way: each `*Action` must have the `__call__` method, which is used for processing the option it's handling. One of its arguments is the "parser" it is used within. We can make the `kcidb.db.ArgumentParser` (to be `kcidb.db.argparse.ArgumentParser` after your change) accept a driver dictionary as one of its constructor arguments (in addition to already-accepted `database`) and store it in the instance.

Then we can make the `DBHelpAction.__call__()` method assert (using the `assert` statement) that the parser it's being given is an instance of that `ArgumentParser`, and get the driver list from it.

How about this?

**Comment by spbnick** [comment]
> should PatternHelpAction be also be moved to argparse?

Yes, because it's closely related to `argparse` purpose.

**Comment by shivam-Purohit** [comment]
@spbnick I tried all related changes in one commit. I am getting flake8 and pylinter errors. I will fix all that once all the changes are done. I am also checking for any code errors I made. I have one suggestion if we could add pre-commit check feature for the commits? That will ensure a proper code even before committing.

**Comment by spbnick** [comment]
>  I have one suggestion if we could add pre-commit check feature for the commits? That will ensure a proper code even before committing.

Of course! I have pre-commit hook set up in my local repo. However, I'm not sure how to best distribute such a script. If you have an idea, please post a PR. You can take the commands to run from the GitHub test workflow, in the repo.

**Comment by shivam-Purohit** [comment]
> If you have an idea, please post a PR. You can take the commands to run from the GitHub test workflow, in the repo.

I will surely try that cause this way is really trying. I have a bunch of more ideas in my mind I will message in the slack

**Comment by shivam-Purohit** [comment]
ERROR STATEMENT
```
ImportError while loading conftest '/home/shivam/Desktop/outreachy/kcidb/conftest.py'.
conftest.py:7: in <module>
    import kcidb.db
kcidb/__init__.py:6: in <module>
    from kcidb.db.argparse import QueryArgumentParser
kcidb/db/__init__.py:9: in <module>
    from kcidb.db.argparse import QueryArgumentParser
kcidb/db/argparse.py:103: in <module>
    class OutputArgumentParser(kcidb.argparse.OutputArgumentParser):
E   AttributeError: partially initialized module 'kcidb' has no attribute 'argparse' (most likely due to a circular import)
```

**Comment by spbnick** [comment]
Sorry, I wasn't able to reproduce your problem. Please push the exact code you have problem with to this PR and specify how to reproduce the issue.

File: kcidb/db/argparse.py
Comment (shivam-Purohit) [inline kcidb/db/argparse.py:27]: This is the file I am having an error.  See we are importing both argparse and kcidb/argparse. That creates confusion.

Comment (shivam-Purohit) [reply kcidb/db/argparse.py:27]: You have only one argparse in your package where I have multiple. When I import the argparse  This can be solved if we import argparse as ap(or something else). So then it does not confuse with other argparse that we will import from kcidb.

**Comment by shivam-Purohit** [comment]
```

class ArgumentParser(kcidb.argparse.ArgumentParser):
    """
    Command-line argument parser with common database arguments added.
    """

    def __init__(self, driver_types, *args, database=None, **kwargs):
        """
        Initialize the parser, adding common database arguments.

        Args:
            args:       Positional arguments to initialize ArgumentParser
                        with.
            database:   The default database specification to use, or None to
                        make database specification required.
            kwargs:     Keyword arguments to initialize ArgumentParser with.
        """
        self.DRIVER_TYPES = driver_types
        super().__init__(*args, **kwargs)
        add_args(self, database=database)
```
This is the ArgumentParser we defined in kcidb/db/argparse where constructor takes driver_types.
This db.ArgumentParser is also called in kcidb/argparse. 
here we haven't passed driver_types are parameter.
 ```
parser = db.ArgumentParser(database="sqlite::memory:",
                               description=description)
```
This might be the reason for the error?

**Comment by spbnick** [comment]
> This is the ArgumentParser we defined in kcidb/db/argparse where constructor takes driver_types.
> This db.ArgumentParser is also called in kcidb/argparse.
> here we haven't passed driver_types are parameter.

> parser = db.ArgumentParser(database="sqlite::memory:", description=description)

> This might be the reason for the error?

I doubt it. I have a hunch about what's going on, and I appreciate you trying to figure it out. However, if you want my help, please explain exactly how I can reproduce your problem, with the code in your PR. Thank you!

**Comment by shivam-Purohit** [comment]
@spbnick The way to reproduce will be then you will have to separate argparse in kcidb and kcidb/db. In db we are calling kcidb.argparse and in kcidb,argparse we are import that db.argparse.ArgumentParser. That may be the reason of circular import.

**Comment by spbnick** [comment]
> @spbnick The way to reproduce will be then you will have to separate argparse in kcidb and kcidb/db. In db we are calling kcidb.argparse and in kcidb,argparse we are import that db.argparse.ArgumentParser. That may be the reason of circular import.
Before we weren't facing this issue because we had these functions in the third module mainly misc or init.py

Could you push the exact code you have the problem with to this (or another PR), and give me the exact command to execute to reproduce this? If this is still a problem, let's have a video call tomorrow, and go over this. Reach out to me on Slack and tell me the possible times for a call, if you'd like to do that.

> PS: unless I finish this I cannot work on other as it is not allowing me to change branches ( i have build folder that is not committed)

Even though VS code could be making switching branches difficult, you could always clone the repository in another directory and work on another branch there.

**Comment by shivam-Purohit** [comment]
@spbnick Can you review?

File: kcidb/db/__init__.py
Comment (spbnick) [inline kcidb/db/__init__.py:12]: Could you avoid importing this separately, and just use `argparse.QueryArgumentParser`?

File: kcidb/db/argparse.py
Comment (spbnick) [inline kcidb/db/argparse.py:92]: Could you add an assertion here verifying that `driver_types` is a dictionary of strings and types?

File: kcidb/db/argparse.py
Comment (spbnick) [inline kcidb/db/argparse.py:86]: Please document the `driver_types` argument here.

File: kcidb/db/argparse.py
Comment (spbnick) [inline kcidb/db/argparse.py:98]: We need **all** database argument parsers to have the `driver_types` attribute, implemented similar to the `ArgumentParser` above. Otherwise the `--database-help` option usage would fail for commands using them. Could you do that?

File: kcidb/misc.py
Comment (spbnick) [inline kcidb/misc.py:3]: Please do not post commented-out code, here, and below. Just remove it.

File: kcidb/misc.py
Comment (spbnick) [inline kcidb/misc.py:15]: Please do not do unrelated whitespace/formatting changes. If you need to change this for some reason, please make a separate commit with this change explaining the reasoning in the commit message.

File: kcidb/monitor/spool/__init__.py
Comment (spbnick) [inline kcidb/monitor/spool/__init__.py:16]: Actually, `iso_timestamp` needs to be in `kcidb.argparse` as well, since it's acting (and is documented) as an argparse "argument type".

File: kcidb/mq/__init__.py
Comment (spbnick) [inline kcidb/mq/__init__.py:17]: Could we just do `import kcidb.argparse` and refer to `output_add_args` as `kcidb.argparse.output_add_args`?

File: kcidb/mq/__init__.py
Comment (spbnick) [inline kcidb/mq/__init__.py:21]: Could we do `from kcidb.mq import argparse` instead, and use `argparse.SubscriberArgumentParser` and so on, instead?

File: kcidb/mq/argparse.py
Comment (spbnick) [inline kcidb/mq/argparse.py:1]: Could you expand this a bit? To something like "KCIDB argument parsing"? Applies to similar docstrings in other modules too. Also, typo in "KCIDB".

File: kcidb/oo/__init__.py
Comment (spbnick) [inline kcidb/oo/__init__.py:13]: We don't do relative imports in general, let's not introduce them piecemeal, and keep using the absolute imports for now. If we're going to change that, then it's better to come up with a general scheme and change everything at once. Uniformity makes code easier to read.

**Review by spbnick — CHANGES_REQUESTED** [review]
This is starting to take shape!

I left a bunch of comments/requests inline. Apart from them, please make sure your commits are [in order](https://kernelci.org/docs/kcidb/developer_guide/#commits), i.e. change-atomic and described appropriately, before review. This change seems to fit a single commit.

Thank you :heart:

File: kcidb/orm/__init__.py
Comment (shivam-Purohit) [reply kcidb/orm/__init__.py:160]: Sorry, it might have happened when I was just starting with the issue. I refactored the code

File: kcidb/db/argparse.py
Comment (shivam-Purohit) [reply kcidb/db/argparse.py:93]: I have added the code. That should work I guess. I have to check driver_types as dict as well right?

File: kcidb/__init__.py
Comment (shivam-Purohit) [reply kcidb/__init__.py]: all these imports were the result of me solving the circular import problem. Apparently, these imports weren't the actual problem. I have fixed them in the commit.

File: kcidb/orm/__init__.py
Comment (spbnick) [reply kcidb/orm/__init__.py]: No problem. The general guidance with code changes, either just for yourself, or preparing for review is to change only what's necessary for the task. It's good practice to check the diffs before posting a PR/requesting a review to make sure that nothing extra is changed.

**Comment by shivam-Purohit** [comment]
@spbnick do we need `logging_setup` in kcidb/misc. I think I moved it to argparse.  I tried reviewing my pr. Got a few unnecessary changes. I have added a second commit. I think I will need to club commits once I am done with the changes. Can you review it?

File: kcidb/db/argparse.py
Comment (spbnick) [inline kcidb/db/argparse.py:144]: Could you align those descriptions on the same column, same as was before your change, but now further to the right? Re-wrap them as necessary to match the column limit, please.

**Review by spbnick — CHANGES_REQUESTED** [review]
The code seems to look pretty good, but the inclusion of unnecessary formatting changes makes it difficult to review, so I didn't look too hard.

I left a few comments inline, but I think we need to split this change into a few commits, to make it easier to manage, in this order:
1. Make the kcidb.db argument parsers accept the `driver_types` dictionary, make the help action use it, and the tools pass the dictionary to the parsers.
2. Make the kcidb.orm argument parsers accept the `Pattern.STRING_DOC` value as an argument (via a parameter named `string_doc`, for example), and store it (not the `Pattern` class itself, as you started writing it), similarly to how `driver_types` is stored in `kcidb.db`. And do the same thing with the tools supplying it. Alternatively, wait for #396 to get merged and just refer to the moved `kcidb.orm.query.Pattern` symbol directly, without passing it around through the parsers.
3. Move the resulting argparse code.

Make sure that you only have those three commits, and don't change code/formatting/whitespace unnecessarily.

Thank you for you hard work!

**Comment by spbnick** [comment]
Hope your exams went well, @shivam-Purohit!

**Comment by shivam-Purohit** [comment]
@spbnick I was about to ask you to review hahaha. Can you? I will work on those on weekend.

**Comment by spbnick** [comment]
> @spbnick I was about to ask you to review hahaha. Can you? I will work on those on weekend.

I just did. I think that's gonna be plenty of work for the weekend :see_no_evil: But send me a message if you run out :grin:

**Comment by shivam-Purohit** [comment]
> I just did. I think that's gonna be plenty of work for the weekend 🙈 But send me a message if you run out 😁

no that's okay I will have plenty of time this week

File: kcidb/misc.py
Comment (shivam-Purohit) [reply kcidb/misc.py:86]: how should I identify any other that needs to be moved?

**Comment by shivam-Purohit** [comment]
> 1. Make the kcidb.db argument parsers accept the `driver_types` dictionary, make the help action use it, and the tools pass the dictionary to the parsers.

isn't it done already?

**Comment by shivam-Purohit** [comment]
> 2\. And do the same thing with the tools supplying it.

what are the tools supplying it? like the ArgumentParser it is inheriting?

**Comment by spbnick** [comment]
>> 2. And do the same thing with the tools supplying it.

> what are the tools supplying it? like the ArgumentParser it is inheriting?

The tools which have support for outputting the Pattern.STRING_DOC. That is those which (eventually) call `kcidb.orm.arparse_add_args()`. Those would be implemented by `kcidb.oo.query_main` and `kcidb.notify_main`. And perhaps nothing else.

However, I think it would be much better for @octonawish-akcodes's PR to get merged first, as otherwise it's too much churn: we would likely end up removing this once the `orm` module is broken down.

Thinking about it, perhaps we could do better than having the `driver_types` passed to all those argument parsers as well (store the `DRIVER_DICT` in a `kcidb.db` submodule?). But let's not make it even longer and keep the present course for now.

**Comment by spbnick** [comment]
I'll take a look at it again, once #396 is merged. Meanwhile, could you please organize the commits, i.e. split into two commits (now that we're waiting for that PR and don't handle `Pattern.STRING_DOC`): one handling the `driver_dict` transition first, and another extracting the argparse modules (the main task)?

**Comment by shivam-Purohit** [comment]
> I'll take a look at it again, once #396 is merged. Meanwhile, could you please organize the commits, i.e. split into two commits (now that we're waiting for that PR and don't handle `Pattern.STRING_DOC`): one handling the `driver_dict` transition first, and another extracting the argparse modules (the main task)?

sure

**Comment by shivam-Purohit** [comment]
I don't know why the rebase changes my code (move back in time) it should be keeping the code the same and only merge commit into one. I will redo the changes and add separate commitsfor driver_types

**Comment by shivam-Purohit** [comment]
this is the error I was getting in the pytest. I could not trace the error. Could you help me with that @spbnick 
```

Traceback (most recent call last):
  File "/home/shivam/Desktop/ahhhh/kcidb/test_kcidb.py", line 105, in test_query_main
    assert_executes("", *argv, driver_source=driver_source)
  File "/home/shivam/Desktop/ahhhh/kcidb/kcidb/unittest.py", line 90, in assert_executes
    raise AssertionError("\n".join(errors))
AssertionError: Expected exit status 0, got 1
Stderr doesn't match regex '':
    AttributeError: 'Namespace' object has no attribute 'issue_ids
```

**Comment by octonawish-akcodes** [comment]
Your changes didn't have the `data.py` file in the `orm` module 🤔

**Comment by shivam-Purohit** [comment]
@spbnick I did it through the second method you told me on call. Can you review it?

File: kcidb/argparse.py
Comment (spbnick) [inline kcidb/argparse.py:31]: These `LOGGING_*` symbols shouldn't be here, and they definitely shouldn't me simply copied here. They should be defined in one place only, to avoid desynchronization. I see that both the `kcidb.misc` and `kcidb.argparse` modules use those. To avoid both this copying and a circular import, we need to move all symbols starting with `LOGGING_` or `logging_` to their own module called `kcidb.logging` (in `kcidb/logging.py` file).

After moving the symbols should naturally lose their `LOGGING_`/`logging_` prefixes, so e.g. `kcidb.misc.LOGGING_LEVEL_MAP` becomes `kcidb.logging.LEVEL_MAP`, and `kcidb.misc.logging_setup` becomes `kcidb.logging.setup`, and so on.

This change would better be done **before** extracting the argparse modules, in a separate commit.

File: kcidb/db/__init__.py
Comment (spbnick) [inline kcidb/db/__init__.py]: GitHub still doesn't allow commenting on unchanged lines in PRs, but this comment belongs to line 407:

This commit (and not the next one as it is right now) needs to also make DBHelpAction actually use the `driver_types` from the parsers (instead of using the global DRIVER_TYPES). Otherwise this change is incomplete, and the classes are not ready for extraction into their own `argparse` module.

File: kcidb/db/argparse.py
Comment (spbnick) [inline kcidb/db/argparse.py:96]: Please do not introduce unrelated whitespace changes. Please remove this blank line.

**Review by spbnick — CHANGES_REQUESTED** [review]
This is looking much better now! Thank you, Shivam!

Most of the inline comments are cosmetic, but there's one concerning a problem with logging-related symbols which would need another separate commit, and another regarding the completeness of `driver_types` change.

File: kcidb/oo/argparse.py
Comment (spbnick) [inline kcidb/oo/argparse.py:3]: Please remove the extra double quote at the beginning, the extra space before the closing `"""`, and format the docstring the same way as other docstrings are. I.e.:
```suggestion
"""
KCIDB object-oriented (OO) data representation-management
command-line argument parsing
"""
```

File: kcidb/orm/argparse.py
Comment (spbnick) [inline kcidb/orm/argparse.py:58]: Please, revert changing the double quotes to single quotes in this function, in this commit, as it's unrelated.
If you feel that should be changed, do that in a separate commit, and provide a justification in the commit message.

File: kcidb/db/argparse.py
Comment (spbnick) [inline kcidb/db/argparse.py:1]: Please remove the extra double quote at the beginning.

File: kcidb/mq/argparse.py
Comment (spbnick) [inline kcidb/mq/argparse.py:1]: Please remove the extra double quote at the beginning.

File: kcidb/orm/argparse.py
Comment (spbnick) [inline kcidb/orm/argparse.py:3]: Please remove the leading whitespace in the last line and format the docstring similar to the others. I.e.:
```suggestion
"""
KCIDB object-relational mapping (ORM)-management
command-line argument parsing
"""
```

**Review by spbnick — CHANGES_REQUESTED** [review]
Thank you, Shivam. I only have a bunch of minor requests. Otherwise this looks great.

Also, please "resolve conversation" for the requests that you implemented. Otherwise we'll be getting lost in them :see_no_evil: 

Thanks!

File: kcidb/__init__.py
Comment (shivam-Purohit) [reply kcidb/__init__.py:8]: should i remove the` # noqa` now?

File: kcidb/orm/argparse.py
Comment (spbnick) [reply kcidb/orm/argparse.py:59]: Thank you! That space is there to separate the first sentence from the second. So please keep it.

File: kcidb/__init__.py
Comment (spbnick) [reply kcidb/__init__.py]: We need the `# noqa` for flake8 to ignore that we don't necessarily use the imported modules - they're needed to make life easier for the package users (what #404 is talking about). So, please don't remove it (that would likely make CI fail anyway). We could use a more specific waiver, though: `# noqa: F401`. That would be nice to do, but it would have to be done in a separate commit (maybe updating all similar `# noqa` comments).

**Comment by shivam-Purohit** [comment]
@spbnick I have made the changes. 
If it's okay we can work on `# noqa`.  The commit should come after the separation of argparse right as it is unrelated to argparse change.

**Comment by spbnick** [comment]
> If it's okay we can work on # noqa.

Great!

> The commit should come after the separation of argparse right as it is unrelated to argparse change.

Doesn't really matter, as they're unrelated. Pick the order that's easier and creates less churn.

**Comment by shivam-Purohit** [comment]
@spbnick how bout this?

File: kcidb/orm/argparse.py
Comment (spbnick) [inline kcidb/orm/argparse.py:49]: Please revert this change of double quotes to single quotes:
```suggestion
        nargs="*",
```
It's unrelated to this commit.

File: kcidb/orm/argparse.py
Comment (spbnick) [inline kcidb/orm/argparse.py:53]: Same here, please revert this formatting change, and restore the comma on the end, in this commit:
```suggestion
        "See pattern documentation with --pattern-help.",
```
Please avoid making unrelated changes, especially when moving code around. It complicates reviews.

File: kcidb/orm/argparse.py
Comment (spbnick) [inline kcidb/orm/argparse.py:58]: Similarly here:
```suggestion
        help="Print pattern string documentation and exit.",
```

**Review by spbnick — CHANGES_REQUESTED** [review]
Thank you, Shivam! We're almost there!

I found one more piece of unnecessarily-copied code, and three last stray changes, that need reverting.

To catch stray changes like that you could use a diff tool (e.g. `diff` itself, `vimdiff`, or something with a GUI). Get the file before your code move, and compare it to the file where you moved the code. E.g. like this:

```bash
vimdiff <(git show HEAD^^:kcidb/orm/__init__.py) kcidb/orm/argparse.py
```

**Pull request closed without merging**
